src/wallet/home_cards/deferred_deep_link/parser.rs

```diff
@@ -0,0 +1,286 @@
+    fn decode(
+        &self,
+        encoded_value: impl AsRef<str>,
+    ) -> Result<OnboardingDeepLinkValue> {
+        let decoded_value_json_bytes = URL_SAFE_NO_PAD
+            .decode(encoded_value.as_ref())
+            .map_err(|e| {
+                println!("{}", e);
+                CommonError::DeferredDeepLinkInvalidValueFormat
+            })?;
+
+        serde_json::from_slice::<OnboardingDeepLinkValue>(
+            decoded_value_json_bytes.as_ref(),
+        )
+        .map_err(|_| CommonError::DeferredDeepLinkDecodingFailed)
+    }
+}
+
+#[cfg(test)]
+mod tests_decode {
+    use super::*;
+
+    #[allow(clippy::upper_case_acronyms)]
+    type SUT = Parser;
+
+    fn make_sut() -> SUT {
+        SUT::new(GatewayClient::new(
+            Arc::new(MockAntenna::new_always_failing()),
+            NetworkID::Stokenet,
+        ))
+    }
+
+    #[test]
+    fn decode_correct() {
+        let sut = make_sut();
+        let encoded_value = "eyJtZXRob2QiOiJtb2JpbGUiLCJkYXBwX3JlZmVycmVyIjoiYWNjb3VudF9yZHgxMjh5Nmo3OG10MGFxdjYzNzJldnoyOGhyeHA4bW4wNmNjZGRrcjd4cHBjODhoeXZ5bnZqZHdyIiwic3BlY2lhbF9kYXBwIjoicmFkcXVlc3QifQ";
+        let result = sut.decode(encoded_value).unwrap();
+        assert_eq!(result, OnboardingDeepLinkValue::sample());
+    }
+
+    #[test]
+    fn decode_invalid_value_format() {
+        let sut = make_sut();
+        let encoded_value = "invalid format";
+        let result = sut.decode(encoded_value).unwrap_err();
+        assert_eq!(result, CommonError::DeferredDeepLinkInvalidValueFormat);
+    }
+
+    #[test]
+    fn decode_decoding_failed() {
+        let sut = make_sut();
+        let encoded_value = "e30";
+        let result = sut.decode(encoded_value).unwrap_err();
+        assert_eq!(result, CommonError::DeferredDeepLinkDecodingFailed);
+    }
+}
+
+impl Parser {
+    async fn transform_onboarding_deep_link_value(
+        &self,
+        value: OnboardingDeepLinkValue,
+    ) -> Result<HomeCards> {
+        let mut result = Vec::new();
+
+        if let Some(special_dapp) = value.special_dapp {
+            if special_dapp == DeferredDeepLinkSpecialDapp::RadQuest {
+                result.push(HomeCard::ContinueRadQuest);
+            } else {
+                result.push(HomeCard::StartRadQuest);
+            }
+        } else {
+            result.push(HomeCard::StartRadQuest);
+        }
+
+        if let Some(dapp_referrer) = value.dapp_referrer {
+            match self.gateway_client.fetch_dapp_metadata(dapp_referrer).await {
+                Ok(metadata) => {
+                    if let Some(icon_url) = metadata.get_icon_url() {
+                        result.push(HomeCard::Dapp {
+                            icon_url: Some(icon_url),
+                        });
+                    } else {
+                        result.push(HomeCard::Dapp { icon_url: None });
+                    }
+                }
+                Err(_) => result.push(HomeCard::Dapp { icon_url: None }),
+            }
```

Comment:
You should be able to convert this, to something like this:
```rust
let icon_url = self.gateway_client.fetch_dapp_metadata(dapp_referrer).await
        .ok()
        .and_then(|metadata| metadata.get_icon_url());
    
result.push(HomeCard::Dapp { icon_url });
```
given that HomeCard::Dapp is always appened in the all execution paths.

Reply:
nice x2!